src/sync/pull.rs

```diff
@@ -490,4 +502,628 @@ mod tests {
+
+                assert_eq!(&sync_head_hash, &got_resp.as_ref().unwrap().sync_head);
+            } else {
+                let got_head = read.get_head(SYNC_HEAD_NAME).await.unwrap();
+                assert!(
+                    got_head.is_none(),
+                    format!(
+                        "{}: expected head to be None, was {}",
+                        c.name,
+                        got_head.unwrap()
+                    )
+                );
+                // In a nop sync we except Beginpull to succeed but sync_head will
+                // be empty.
+                if c.exp_err.is_none() {
+                    assert!(&got_resp.as_ref().unwrap().sync_head.is_empty());
+                }
+            }
+
+            // Check that BeginTryPullResponse is filled like we would expect.
+            if c.exp_err.is_none() {
+                if let Ok(pull_response) = &c.pull_result {
+                    assert_eq!(
+                        pull_response.client_view_info,
+                        got_resp.unwrap().client_view_info
+                    );
+                }
+            }
+        }
+    }
+
+    pub struct FakePuller<'a> {
+        exp_pull_req: &'a PullRequest,
+        exp_diff_server_url: &'a str,
+        exp_diff_server_auth: &'a str,
+        exp_sync_id: &'a str,
+
+        // We would like to write here:
+        //    result: Result<PullResponse, PullError>,
+        // but pull takes &self so we can't move out of result if we did.
+        // Cloning and returning result would work except for that our error
+        // enums contain values that are not cloneable, eg http::Status and
+        // DeserializeErr. (Or, I guess we could make pull take &mut self as another
+        // solution, so long as all contained errors are Send. I think.)
+        resp: Option<PullResponse>,
+        err: Option<String>,
+    }
+
+    #[async_trait(?Send)]
+    impl<'a> Puller for FakePuller<'a> {
+        async fn pull(
+            &self,
+            pull_req: &PullRequest,
+            diff_server_url: &str,
+            diff_server_auth: &str,
+            sync_id: &str,
+        ) -> Result<PullResponse, PullError> {
+            assert_eq!(self.exp_pull_req, pull_req);
+            assert_eq!(self.exp_diff_server_url, diff_server_url);
+            assert_eq!(self.exp_diff_server_auth, diff_server_auth);
+            assert_eq!(self.exp_sync_id, sync_id);
+
+            match &self.err {
+                Some(s) => match s.as_str() {
+                    "FetchNotOk(500)" => Err(PullError::FetchNotOk(
+                        http::StatusCode::INTERNAL_SERVER_ERROR,
+                    )),
+                    _ => panic!("not implemented"),
+                },
+                None => {
+                    let r = self.resp.as_ref();
+                    Ok(r.unwrap().clone())
+                }
+            }
+        }
+    }
+
+    #[async_std::test]
+    async fn test_maybe_end_pull() {
```

Comment:
think this should be test_maybe_end_try_pull and the function should be maybe_end_try_pull.